js/serializers/scormSuspendDataSerializer.js

```diff
@@ -1,691 +1,1129 @@
+     * @returns {undefined|Error}
+     */
+    getInvalidTypeError(value) {
+      try {
+        value = convertValuesToNumbers(value);
+        if (esTypeOf(value) === 'array') {
+          findValueTypeFromValues(value);
         } else {
-          var hasVariableLengthChildren = arrayItemType.size == "variable";
-          if (hasVariableLengthChildren) {
-            var VLDS = DataType.VARIABLELENGTHDESCRIPTORSIZE;
-            while ( binItems != "" ) {
-
-              var variableLength = Converter._variableLength( binItems );
-              var binItem = binItems.substr(0, VLDS + variableLength);
-              binItems = binItems.substr(VLDS+variableLength);
-              //console.log("binItem", binItem, BinaryToNumber(binItem, 16));
-
-              rtn.push( Converter._converters['binary'][ arrayItemType.name ]( binItem) );
-            }
-          } else {
-            while ( binItems != "" ) {
-              var binItem = binItems.substr(0, arrayItemType.length);
-              binItems = binItems.substr(arrayItemType.length);
-
-              rtn.push( Converter._converters['binary'][ arrayItemType.name ](binItem) );
-            }
-          }
-
+          findValueTypeFromValue(value);
         }
+        return;
+      } catch(err) {
+        return err;
+      }
+    }
 
-
-        return rtn;
-
-      },
-      "base64": function ConverterBinaryToBase64(bin) { //TODO PADDING NOT GOOD
-        var paddingLength = 0;
-        if (bin.length % 6) paddingLength = 6 - (bin.length % 6);
-        binPaddingLen = NumberToBinary(paddingLength, 6);
-        binPadding = Padding.addLeft("", paddingLength);
-        bin = binPaddingLen + binPadding + bin;
-
-        var binLength = bin.length;
-        var base64 = "";
-        for (var b = 0; b < 10000; b++) {
-          if (b*6 >= binLength) break;
-
-          var block = bin.substr(b*6,6);
-          base64 += Base64(parseInt(block, 2));
+    /**
+     * Convert an array, boolean or number into a string binary representation
+     * @param {number|boolean|Array} value
+     * @param {string} [typeName] To force an internal data type for testing
+     * @param {boolean} [logStats] Console logs statistics on type use
+     * @returns {string}
+     */
+    valueToBin(value, typeName = null, logStats = null) {
+      this.clearLog();
+      const hasInitialType = (typeof typeName === 'string');
+      value = convertValuesToNumbers(value);
+      let valueType;
+      if (hasInitialType) {
+        valueType = findValueTypeFromName(typeName);
+        if (!valueType) {
+          throw new Error(`Could not find value type from name "${typeName}" accepts: ${Object.keys(ValueTypes.nameIndex)}. Leave empty for auto-detect.`);
+        }
+      } else {
+        valueType = findValueTypeFromValue(value);
+        if (!valueType) {
+          throw new Error(`Could not find value type from value "${value}" accepts: ${Object.keys(ValueTypes.nameIndex)}. Leave empty for auto-detect.`);
```

Comment:
Please could we reduce the duplication/nesting:
```suggestion
      const valueType = hasInitialType ?
        findValueTypeFromName(typeName) :
        findValueTypeFromValue(value);
      if (!valueType) {
        throw new Error(`Could not find value type from name "${hasInitialType ? typeName : value}" accepts: ${Object.keys(ValueTypes.nameIndex)}. Leave empty for auto-detect.`);
```

Reply:
yup

Reply:
Did you know you can nest template strings? It makes the error string less meaty:
```js
throw new Error(`Could not find value type from ${hasInitialType ? `name "${typeName}"` : `value "${value}"`} accepts: ${Object.keys(ValueTypes.nameIndex)}. Leave typeName empty for auto-detect.`);
```

Reply:
Can you give me an example?